stan/math/opencl/kernels/matrix_multiply.hpp

```diff
@@ -114,6 +114,94 @@ const local_range_kernel<cl::Buffer, cl::Buffer, cl::Buffer, int, int, int,
     matrix_multiply("matrix_multiply", matrix_multiply_kernel_code,
                     {{"THREAD_BLOCK_SIZE", 32}, {"WORK_PER_THREAD", 8}});
 
+// \cond
+static const char* matrix_vector_multiply_kernel_code = STRINGIFY(
+    // \endcond
+    /**
+     * Matrix-vector multiplication R=A*B on the OpenCL device
+     *
+     * @param[in] A matrix in matrix-vector multiplication
+     * @param[in] B vector in matrix-vector multiplication
+     * @param[out] R the output vector
+     * @param[in] M Number of rows for matrix A
+     * @param[in] N Number of cols for matrix A and number of rows for vector B
+     */
+    __kernel void matrix_vector_multiply(
+        const __global double* A, const __global double* B, __global double* R,
+        const int M, const int N) {
+      const int gid = get_global_id(0);
+
+      double acc = 0;
+      for (int i = 0, j = 0; i < N; i++, j += M) {
+        acc += A[j + gid] * B[i];
+      }
+      R[gid] = acc;
+    }
+    // \cond
+);
+// \endcond
+
+/**
+ * See the docs for \link kernels/matrix_multiply.hpp matrix_vector_multiply()
+ * \endlink
+ */
+const global_range_kernel<cl::Buffer, cl::Buffer, cl::Buffer, int, int>
+    matrix_vector_multiply("matrix_vector_multiply",
+                           matrix_vector_multiply_kernel_code);
+
+// \cond
+static const char* row_vector_matrix_multiply_kernel_code = STRINGIFY(
+    // \endcond
+    /**
+     * Row vector-matrix multiplication R=A*B on the OpenCL device
+     *
+     * @param[in] A row vector in row vector-matrix multiplication
+     * @param[in] B matrix in row vector-matrix multiplication
+     * @param[out] R the output vector
+     * @param[in] N Number of cols for vector A and number of rows for matrix B
+     * @param[in] K Number of cols for matrix B
+     */
+    __kernel void row_vector_matrix_multiply(
+        const __global double* A, const __global double* B, __global double* R,
+        const int N, const int K) {
+      const int lid = get_local_id(0);
+      const int gid = get_global_id(0);
+      const int wgid = get_group_id(0);
+
+      double acc = 0;
+      for (int i = lid; i < N; i += 64) {
```

Comment:
64 should be a constant like THREAD_BLOCK_SIZE for matrix multiply. Also add a constant that defines that we are then doing subsequent reduces of size 4.

Reply:
Do we really need a constant for reduce size? Where would it be used?